lib/Conversion/XTenNNToLinalg.cpp

```diff
@@ -0,0 +1,177 @@
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
+
+namespace xilinx::xten {
+#define GEN_PASS_DECL_CONVERTXTENNNTOLINALG
+#define GEN_PASS_DEF_CONVERTXTENNNTOLINALG
+#include "xten/Conversion/Passes.h.inc"
+} // namespace xilinx::xten
+
+using namespace mlir;
+using namespace amd::xten_nn;
+
+namespace {
+
+int64_t getRank(Value v) {
+  return cast<ShapedType>(v.getType()).getRank();
+}
+
+int64_t getMaxRank(ValueRange operands) {
+  int64_t maxRank = 0;
+  for (Value operand : operands) {
+    maxRank = std::max(maxRank, getRank(operand));
+  }
+  return maxRank;
+}
+
+bool isScalar(Value v) {
+  return getRank(v) == 0;
+}
+
+Value getEmptyTensor(OpBuilder &b, Location loc, ShapedType type,
+                     ArrayRef<Value> dynSizes) {
+  return b.create<tensor::EmptyOp>(loc, type.getShape(), type.getElementType(),
+                                   dynSizes,
+                                   type.cast<RankedTensorType>().getEncoding());
+}
+
+inline Value getConstantOrSplat(OpBuilder *b, Location loc, Type t,
+                                Attribute v) {
+  if (VectorType vecType = t.dyn_cast<VectorType>()) {
+    v = SplatElementsAttr::get(vecType, v);
+  }
+  return b->create<arith::ConstantOp>(loc, t, cast<TypedAttr>(v));
+}
+
+Value mapSignOpToStdScalarOp(Location loc, ArrayRef<Type> resultTypes,
+                             Value operand, OpBuilder *b) {
+  Type elementType = getElementTypeOrSelf(operand.getType());
+  if (auto floatType = elementType.dyn_cast<FloatType>()) {
+    Value zero =
+        b->create<arith::ConstantOp>(loc, b->getZeroAttr(operand.getType()));
+    Value ne0I1 = b->create<::mlir::arith::CmpFOp>(
+        loc, arith::CmpFPredicate::ONE, operand, zero);
+    Value ne0Float =
+        b->create<::mlir::arith::UIToFPOp>(loc, zero.getType(), ne0I1);
+    Value copySign = b->create<::mlir::math::CopySignOp>(loc, resultTypes,
+                                                         ne0Float, operand);
+    auto isNan = b->create<::mlir::arith::CmpFOp>(
+        loc, arith::CmpFPredicate::UNO, operand, operand);
+    return b->create<::mlir::arith::SelectOp>(loc, isNan, operand, copySign);
+  }
+  if (auto integerType = elementType.dyn_cast<IntegerType>()) {
+    // sign(x) = x == 0 ? 0 : ((x s>> 31) | 1)
+    Value zero =
+        b->create<arith::ConstantOp>(loc, b->getZeroAttr(operand.getType()));
+    Value bitwidthMinusOne = getConstantOrSplat(
+        b, loc, operand.getType(),
+        b->getIntegerAttr(integerType, integerType.getWidth() - 1));
+    Value one = getConstantOrSplat(b, loc, operand.getType(),
+                                   b->getIntegerAttr(integerType, 1));
+    Value cmp = b->create<::mlir::arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                                 operand, zero);
+    Value ashr =
```

Comment:
Now I get what you are doing. Could you please add a comment to explain it to others?